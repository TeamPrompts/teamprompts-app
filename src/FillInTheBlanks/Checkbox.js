import React from 'react';

function Checkbox({ id, mode, onChange, value }) {
  const inputId = `${id}-${value}`;
  return (
    <label htmlFor={inputId} className="cursor-pointer">
      <div className="relative">
        <input
          checked={mode === value}
          data-testid={inputId}
          id={inputId}
          name={id}
          onChange={onChange}
          type="checkbox"
          value={value}
          className="hidden"
        />

        <div className="toggle__line w-6 h-3 bg-gray-400 rounded-full shadow-inner" />

        <div className="toggle__dot absolute w-4 h-4 bg-gray-300 rounded-full shadow inset-y-0 left-0" />
      </div>
    </label>
  );
}

export default Checkbox;
