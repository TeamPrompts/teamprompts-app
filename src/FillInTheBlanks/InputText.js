import classname from 'classnames';
import React, { useState } from 'react';
import { BLANK } from '../constants';

function InputText({ hint }) {
  const [dirty, setDirty] = useState(false);
  const [size, setSize] = useState(hint.length);

  function onChange(value) {
    setDirty(true);
    if (value.length > hint.length) {
      setSize(value.length);
    } else {
      setSize(hint.length);
    }
  }

  return (
    <span className="inline-flex flex-col leading-none">
      <input
        className={classname({ 'bg-purple-100': dirty })}
        data-testid="input"
        onChange={e => onChange(e.target.value)}
        placeholder={BLANK}
        size={size}
        type="text"
      />
      {hint !== BLANK && <label className="self-center text-sm">{hint}</label>}
    </span>
  );
}

export default InputText;
