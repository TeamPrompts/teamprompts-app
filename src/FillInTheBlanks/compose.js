import { SEPARATOR } from '../constants';

function compose({ build, source, values }) {
  const text = source.slice();
  const parts = text.split(SEPARATOR);
  const result = [];
  parts.forEach((part, index) => {
    result.push(part);
    if (values[index]) {
      result.push(build({ index }));
    }
  });
  return result;
}

export default compose;
