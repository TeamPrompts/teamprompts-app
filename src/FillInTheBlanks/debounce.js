// RESOURCES:
// - https://redd.one/blog/debounce-vs-throttle
// - https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/clearTimeout

function debouce(func, duration) {
  let timeoutId;

  return (...args) => {
    function callback() {
      timeoutId = undefined;
      func.apply(this, args);
    }

    clearTimeout(timeoutId);

    timeoutId = setTimeout(callback, duration);
  };
}

export default debouce;
