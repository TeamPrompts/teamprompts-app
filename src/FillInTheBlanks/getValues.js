import { BLANK, BLANK_XS, modes } from '../constants';

function makeBlanks({ blank, source }) {
  return source.match(/__BLANK__/g).map(() => blank); // INFO: not sure if I can interpolate inside a regex
}

function getValues(mode, { examples, prompts, source }) {
  let values = [];
  const blanks = makeBlanks({ blank: BLANK, source });
  switch (mode) {
    case modes.blanks:
    default:
      values = blanks;
      break;
    case modes.blanksXS:
      values = makeBlanks({ blank: BLANK_XS, source });
      break;
    case modes.examples:
      values = examples;
      break;
    case modes.input:
      if (prompts) {
        values = prompts;
      } else {
        values = blanks;
      }
      break;
    case modes.prompts:
      values = prompts;
      break;
  }
  return values;
}

export default getValues;
