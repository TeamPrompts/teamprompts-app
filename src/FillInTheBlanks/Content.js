import classnames from 'classnames';
import React, { useReducer } from 'react';
import InputText from '../components/InputText/InputText';
import { BLANK, SEPARATOR, modes } from '../constants';
import makeInitialState from './makeInitialState';
import reducer from './reducer';

function getValues(mode, { examples, prompts, source }) {
  let values = [];
  const blanks = source.match(/__BLANK__/g).map(() => BLANK); // INFO: not sure if I can interpolate inside a regex
  switch (mode) {
    case modes.blanks:
    default:
      values = blanks;
      break;
    case modes.examples:
      values = examples;
      break;
    case modes.input:
      if (prompts) {
        values = prompts;
      } else {
        values = blanks;
      }
      break;
    case modes.prompts:
      values = prompts;
      break;
  }
  return values;
}

function interpolate({ dispatch, mode, source, state, values }) {
  const text = source.slice();
  const parts = text.split(SEPARATOR);
  const result = [];
  parts.forEach((part, index) => {
    result.push(part);
    if (values[index]) {
      let element;
      if (mode === modes.input) {
        const hint = values[index];
        element = (
          <InputText
            dispatch={dispatch}
            hint={hint}
            id={index}
            key={index}
            value={state[index]}
          />
        );
      } else {
        element = (
          <span
            className={classnames({
              'bg-yellow-200': mode === modes.blanks,
              'bg-green-200': mode === modes.examples,
              'bg-blue-200': mode === modes.prompts
            })}
            key={index}
          >
            {values[index]}
          </span>
        );
      }
      result.push(element);
    }
  });
  return result;
}

function Content({ fitb, mode }) {
  const [state, dispatch] = useReducer(reducer, makeInitialState(fitb.source));
  const values = getValues(mode, fitb);
  return interpolate({ dispatch, mode, source: fitb.source, state, values });
}

export default Content;
