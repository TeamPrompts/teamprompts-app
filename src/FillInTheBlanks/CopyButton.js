import React from 'react';
import { BLANK, modes } from '../constants';
import compose from './compose';
import getValues from './getValues';

function makeBuild({ inputs, mode, values }) {
  function build({ index }) {
    let element;
    if (mode === modes.input) {
      if (inputs && inputs[index]) {
        element = inputs[index];
      } else {
        element = BLANK;
      }
    } else {
      element = values[index];
    }
    return element;
  }
  return build;
}

// TODO: read inputs from the local storage
function CopyButton({ fitb, inputs, mode, onClick }) {
  return (
    <button
      className="bg-blue-500 focus:outline-none font-bold font-mono hover:bg-blue-700 ml-3 px-4 rounded-full text-xs text-white"
      data-testid="copy-button"
      onClick={event => {
        event.stopPropagation();
        const values = getValues(mode, fitb);
        const result = compose({
          build: makeBuild({ inputs, mode, values }),
          source: fitb.source,
          values
        });
        onClick(result.join(''));
      }}
      type="button"
    >
      Copy to clipboard
    </button>
  );
}

export default CopyButton;
