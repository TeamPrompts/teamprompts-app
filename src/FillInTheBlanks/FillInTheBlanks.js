import { faLink } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classnames from 'classnames';
import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { useClipboard } from 'use-clipboard-copy';
import Checkbox, { values } from '../components/Checkbox/Checkbox';
import { modes } from '../constants';
import Content from './Content';
import CopyButton from './CopyButton';

function FillInTheBlanks({ fitb, history, tag }) {
  const clipboard = useClipboard();
  const [inputs, setInputs] = useState();
  const [mode, setMode] = useState(modes.input);

  function onClick() {
    if (history && tag) {
      history.push(`/${tag.slug}/${fitb.id}`);
    }
  }

  return (
    <div
      className={classnames(
        'border-b border-gray-500 leading-loose py-8 text-lg',
        {
          'hover:bg-gray-100 cursor-pointer': history && tag
        }
      )}
      onClick={onClick}
    >
      <Content callback={setInputs} fitb={fitb} mode={mode} />
      {fitb.examples && (
        <Checkbox
          id={fitb.id}
          onChange={value => {
            if (value === values.off) {
              setMode(modes.examples);
            }
            if (value === values.on) {
              setMode(modes.input);
            }
          }}
          value={mode === modes.examples ? values.on : values.off}
        />
      )}
      <CopyButton
        fitb={fitb}
        inputs={inputs}
        mode={mode}
        onClick={value => {
          console.log(value);
          clipboard.copy(value);
        }}
      />
      {false && tag && (
        <Link
          className="ml-3 text-indigo-700 underline"
          to={`/${tag.slug}/${fitb.id}`}
        >
          <FontAwesomeIcon icon={faLink} />
        </Link>
      )}
    </div>
  );
}

export default FillInTheBlanks;
