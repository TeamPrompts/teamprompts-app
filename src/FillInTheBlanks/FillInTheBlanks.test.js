import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';
import { useClipboard } from 'use-clipboard-copy';
import fitb from '../mocks/fitb';
import tag from '../mocks/tag';
import FillInTheBlanks from './FillInTheBlanks';
import { pageTypes } from '../constants';

jest.mock('use-clipboard-copy');

describe('FillInTheBlanks', () => {
  const mockPush = jest.fn();

  const props = {
    fitb,
    history: { push: mockPush },
    tag
  };

  beforeEach(() => {
    mockPush.mockReset();
  });

  it('to match snapshot', () => {
    const tree = create(<FillInTheBlanks {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('toggle', () => {
    const { container, getByTestId } = render(<FillInTheBlanks {...props} />);

    // on
    fireEvent.click(getByTestId(fitb.id));
    expect(container).toMatchSnapshot();

    // off
    fireEvent.click(getByTestId(fitb.id));
    expect(container).toMatchSnapshot();
  });

  it('Copy to clipboard', () => {
    const mockCopy = jest.fn();
    useClipboard.mockImplementation(() => ({ copy: mockCopy }));
    const { getByTestId } = render(<FillInTheBlanks {...props} />);

    fireEvent.click(getByTestId('copy-button'));

    expect(mockCopy).toBeCalled();
  });

  it('navigate to details w/ tag', () => {
    const { getByTestId } = render(<FillInTheBlanks {...props} />);

    fireEvent.click(getByTestId('container'));

    expect(mockPush).toBeCalledWith(`/${tag.slug}/${fitb.id}`);
  });

  it('navigate to details w/o tag', () => {
    const { tag, ...rest } = props;
    const { getByTestId } = render(<FillInTheBlanks {...rest} />);

    fireEvent.click(getByTestId('container'));

    expect(mockPush).toBeCalledWith(`/all/${fitb.id}`);
  });

  it('not enable', () => {
    const { history, ...rest } = props;
    const { getByTestId } = render(<FillInTheBlanks {...rest} />);

    fireEvent.click(getByTestId('container'));

    expect(mockPush).not.toBeCalled();
  });

  it('navigate to category w/ tag', () => {
    const { getByTestId } = render(
      <FillInTheBlanks pageType={pageTypes.PromptsPage} {...props} />
    );

    fireEvent.click(getByTestId('go-to-category'));

    expect(mockPush).toBeCalledWith(`/${tag.slug}`);
  });

  it('navigate to category w/ tag', () => {
    const { tag, ...rest } = props;
    const { getByTestId } = render(
      <FillInTheBlanks pageType={pageTypes.PromptsPage} {...rest} />
    );

    fireEvent.click(getByTestId('go-to-category'));

    expect(mockPush).toBeCalledWith(`/all`);
  });

  it('no history', () => {
    const { history, ...rest } = props;
    const { getByTestId } = render(
      <FillInTheBlanks pageType={pageTypes.PromptsPage} {...rest} />
    );

    fireEvent.click(getByTestId('go-to-category'));

    expect(mockPush).not.toBeCalled();
  });
});
