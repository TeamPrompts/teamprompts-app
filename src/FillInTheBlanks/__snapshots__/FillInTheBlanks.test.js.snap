// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FillInTheBlanks to match snapshot 1`] = `
<div
  className="border-b border-gray-500 leading-loose py-8 text-lg"
  onClick={[Function]}
>
  Do you know the way to 
  <span
    className="inline-flex flex-col leading-none"
  >
    <div
      className="self-center"
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <input
        className=""
        data-testid="input"
        onChange={[Function]}
        placeholder="________________"
        style={
          Object {
            "boxSizing": "content-box",
            "width": "1px",
          }
        }
        value=""
      />
      <div
        style={
          Object {
            "height": 0,
            "left": 0,
            "overflow": "scroll",
            "position": "absolute",
            "top": 0,
            "visibility": "hidden",
            "whiteSpace": "pre",
          }
        }
      >
        
      </div>
      <div
        style={
          Object {
            "height": 0,
            "left": 0,
            "overflow": "scroll",
            "position": "absolute",
            "top": 0,
            "visibility": "hidden",
            "whiteSpace": "pre",
          }
        }
      >
        ________________
      </div>
    </div>
    <label
      className="self-center text-sm"
    >
      place
    </label>
  </span>
  ?
  <span
    className="inline-block ml-3 relative"
  >
    <input
      checked={false}
      className="hidden"
      data-testid="receM4Gus7HH43oLM"
      id="receM4Gus7HH43oLM"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="cursor-pointer"
      htmlFor="receM4Gus7HH43oLM"
    >
      <div
        className="w-6 h-3 bg-gray-400 rounded-full shadow-inner"
      />
      <div
        className="toggle__dot absolute w-4 h-4 bg-gray-300 rounded-full shadow inset-y-0 left-0"
      />
    </label>
  </span>
  <button
    className="bg-blue-500 focus:outline-none font-bold font-mono hover:bg-blue-700 ml-3 px-4 rounded-full text-sm text-white"
    data-testid="copy-button"
    onClick={[Function]}
    type="button"
  >
    Ctrl+C
  </button>
</div>
`;

exports[`FillInTheBlanks toggle 1`] = `
<div>
  <div
    class="border-b border-gray-500 leading-loose py-8 text-lg"
  >
    Do you know the way to 
    <span
      class="bg-green-200"
    >
      San Jose
    </span>
    ?
    <span
      class="inline-block ml-3 relative"
    >
      <input
        class="hidden"
        data-testid="receM4Gus7HH43oLM"
        id="receM4Gus7HH43oLM"
        type="checkbox"
      />
      <label
        class="cursor-pointer"
        for="receM4Gus7HH43oLM"
      >
        <div
          class="w-6 h-3 bg-gray-400 rounded-full shadow-inner"
        />
        <div
          class="toggle__dot absolute w-4 h-4 bg-gray-300 rounded-full shadow inset-y-0 left-0 toggle__dot--checked"
        />
      </label>
    </span>
    <button
      class="bg-blue-500 focus:outline-none font-bold font-mono hover:bg-blue-700 ml-3 px-4 rounded-full text-sm text-white"
      data-testid="copy-button"
      type="button"
    >
      Ctrl+C
    </button>
  </div>
</div>
`;

exports[`FillInTheBlanks toggle 2`] = `
<div>
  <div
    class="border-b border-gray-500 leading-loose py-8 text-lg"
  >
    Do you know the way to 
    <span
      class="inline-flex flex-col leading-none"
    >
      <div
        class="self-center"
        style="display: inline-block;"
      >
        <input
          class=""
          data-testid="input"
          placeholder="________________"
          style="box-sizing: content-box; width: 2px;"
          value=""
        />
        <div
          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
        />
        <div
          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
        >
          ________________
        </div>
      </div>
      <label
        class="self-center text-sm"
      >
        place
      </label>
    </span>
    ?
    <span
      class="inline-block ml-3 relative"
    >
      <input
        class="hidden"
        data-testid="receM4Gus7HH43oLM"
        id="receM4Gus7HH43oLM"
        type="checkbox"
      />
      <label
        class="cursor-pointer"
        for="receM4Gus7HH43oLM"
      >
        <div
          class="w-6 h-3 bg-gray-400 rounded-full shadow-inner"
        />
        <div
          class="toggle__dot absolute w-4 h-4 bg-gray-300 rounded-full shadow inset-y-0 left-0"
        />
      </label>
    </span>
    <button
      class="bg-blue-500 focus:outline-none font-bold font-mono hover:bg-blue-700 ml-3 px-4 rounded-full text-sm text-white"
      data-testid="copy-button"
      type="button"
    >
      Ctrl+C
    </button>
  </div>
</div>
`;
